File: src/Scanner.Core/Checks/StyleUtilities.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace Scanner.Core.Checks;

/// <summary>
/// Helpers for parsing inline style values.
/// </summary>
public static class StyleUtilities
{
    private static readonly Regex FixedLengthRegex = new(
        "^(?<value>-?\\d+(?:\\.\\d+)?)(?<unit>px|pt|pc|cm|mm|in|em|rem)$",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);

    /// <summary>
    /// Gets the last value for a CSS property defined in an inline style string.
    /// </summary>
    /// <param name="style">The style attribute string.</param>
    /// <param name="propertyName">The CSS property name to find.</param>
    /// <returns>The last matching value, or null if not found.</returns>
    public static string? GetLastPropertyValue(string? style, string propertyName)
    {
        if (string.IsNullOrWhiteSpace(style) || string.IsNullOrWhiteSpace(propertyName))
        {
            return null;
        }

        string? value = null;
        var segments = style.Split(';', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
        foreach (var segment in segments)
        {
            var separatorIndex = segment.IndexOf(':');
            if (separatorIndex <= 0)
            {
                continue;
            }

            var name = segment[..separatorIndex].Trim();
            if (!name.Equals(propertyName, StringComparison.OrdinalIgnoreCase))
            {
                continue;
            }

            var propertyValue = segment[(separatorIndex + 1)..].Trim();
            if (!string.IsNullOrWhiteSpace(propertyValue))
            {
                value = propertyValue;
            }
        }

        return value;
    }

    /// <summary>
    /// Determines whether a CSS length value is a fixed size (non-percentage).
    /// </summary>
    /// <param name="value">The CSS length value.</param>
    /// <returns>True when the value is a fixed length; otherwise, false.</returns>
    public static bool IsFixedLength(string value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return false;
        }

        var trimmed = value.Trim();
        if (trimmed.Equals("auto", StringComparison.OrdinalIgnoreCase))
        {
            return false;
        }

        if (trimmed.EndsWith('%'))
        {
            return false;
        }

        var match = FixedLengthRegex.Match(trimmed);
        if (!match.Success)
        {
            return false;
        }

        if (!double.TryParse(match.Groups["value"].Value, NumberStyles.Float, CultureInfo.InvariantCulture, out var number))
        {
            return false;
        }

        return number > 0;
    }
}

File: node_engine/src/checks/StyleUtilities.js
const fixedLengthRegex = /^(?<value>-?\d+(?:\.\d+)?)(?<unit>px|pt|pc|cm|mm|in|em|rem)$/i;

const getLastPropertyValue = (style, propertyName) => {
  if (!style || !propertyName) {
    return null;
  }

  let value = null;
  const segments = style.split(";").map((segment) => segment.trim()).filter(Boolean);
  for (const segment of segments) {
    const separatorIndex = segment.indexOf(":");
    if (separatorIndex <= 0) {
      continue;
    }

    const name = segment.slice(0, separatorIndex).trim();
    if (name.toLowerCase() !== propertyName.toLowerCase()) {
      continue;
    }

    const propertyValue = segment.slice(separatorIndex + 1).trim();
    if (propertyValue) {
      value = propertyValue;
    }
  }

  return value;
};

const isFixedLength = (value) => {
  if (!value || !value.trim()) {
    return false;
  }

  const trimmed = value.trim();
  if (trimmed.toLowerCase() === "auto") {
    return false;
  }

  if (trimmed.endsWith("%")) {
    return false;
  }

  const match = fixedLengthRegex.exec(trimmed);
  if (!match) {
    return false;
  }

  const number = Number.parseFloat(match.groups.value);
  if (Number.isNaN(number)) {
    return false;
  }

  return number > 0;
};

module.exports = { getLastPropertyValue, isFixedLength };
